### DISCUSSION TITLE
"Draft mode" for faster rendering?

### DISCUSSION THREAD

**Discussion by krlmlr** [lvjr/tabularray#273] [Ideas]
We use tabularray with the "long" option a lot, for its superior output quality. However, the rendering is quite slow (>10 seconds for a two-pages document, with longtable it's <1 second).

I wonder if this package could support a "draft" mode that would perhaps omit some of the expensive computations, perhaps at the expense of output quality?

Could there be a "verbose" or "logging" mode that would print detailed information on what tabularray is doing, perhaps with a time stamp?

I saw #250, but this seems a lot of effort, and I'm not sure we can switch to LuaLaTeX today.

**Comment by lvjr** [comment]
> We use tabularray with the "long" option a lot, for its superior output quality. However, the rendering is quite slow (>10 seconds for a two-pages document, with longtable it's <1 second).

It is the cost of quality, because TeX doesn't provide a native data stucture for reading and writing key-value pairs. (Also `longtable` requires multiple compilations but `tabularray` requires only single compilation.)

> I wonder if this package could support a "draft" mode that would perhaps omit some of the expensive computations, perhaps at the expense of output quality?

Possible, but not that easy.

> Could there be a "verbose" or "logging" mode that would print detailed information on what tabularray is doing, perhaps with a time stamp?

There is some tracing information in the log by default:
```
> Step: init table outer spec.
> Step: parse table options.
> Step: split table.
> Step: init table inner spec.
> Step: parse table inner spec.
> Step: execute table commands.
> Step: do functional calculation.
> Step: calculate cell and line sizes.
> Step: build the whole table.
```
And you can print more with `\SetTblrTracing{all}`. You can add time stamps with some hacks. But I don't think ordinary users need these. 

> I saw #250, but this seems a lot of effort, and I'm not sure we can switch to LuaLaTeX today.

I think LuaTeX is the only possible way to go.

PS: please test version 2022C. I believe it is faster than 2022B but not sure how much.

**Comment by lvjr** [comment]
There is a tip for faster rendering: fixing all column sizes.

**Reply by ysalmon** [reply]
Re this, would it be possible for tabularray to output the calculated column width to .aux and use that ? One would juste have to delete the .aux for the final compilation to make the updates that might have become necessary.

**Comment by lvjr** [comment]
I have created a new related issue: https://github.com/lvjr/tabularray/issues/274 .
